### PR TITLE
[SPARK-15029] improve error message for Generate

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This PR improve the error message for `Generate` in 3 cases:
1. generator is nested in expressions, e.g. `SELECT explode(list) + 1 FROM tbl`
2. generator appears more than one time in SELECT, e.g. `SELECT explode(list), explode(list) FROM tbl`
3. generator appears in other operator which is not project, e.g. `SELECT * FROM tbl SORT BY explode(list)`
## How was this patch tested?

new tests in `AnalysisErrorSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -165,6 +167,33 @@ class Analyzer(
...
+      case UnresolvedAlias(_: Generator, _) => false
+      case Alias(_: Generator, _) => false
+      case MultiAlias(_: Generator, _) => false
+      case other if hasGenerator(other) => true
+      case _ => false
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Project(projectList, _) if projectList.exists(hasNestedGenerator) =>
+        throw new AnalysisException("Generators are not supported when it's nested in expressions.")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: i think we need to show the generator itself, as "generator" is a very spark sql specific term. If users just put an explode, he/she wont' know it's a generator.

**Comment by rxin** [comment]
Looks like a legitimate test failure.

**Comment by rxin** [comment]
cc @hvanhovell for review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -1250,40 +1251,77 @@ class Analyzer(
...
-        g.copy(generatorOutput = makeGeneratorOutput(g.generator, g.generatorOutput.map(_.name)))
+      case Project(projectList, _) if projectList.exists(hasNestedGenerator) =>
+        val nestedGenerator = projectList.find(hasNestedGenerator).get
+        throw new AnalysisException("Generators are not supported when it's nested in " +
+          "expressions, but got: " + toPrettySQL(trimAlias(nestedGenerator)))
+
+      case Project(projectList, _) if projectList.count(hasGenerator) > 1 =>
+        val generators = projectList.flatMap(_.find(_.isInstanceOf[Generator]))
+        throw new AnalysisException("Only one generator allowed per select clause but found " +
+          generators.size + ": " + generators.map(toPrettySQL).mkString(", "))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: MINOR: Shouldn't we `trimAlias` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -1250,40 +1251,77 @@ class Analyzer(
...
+      case Alias(child, _) => child
+      case MultiAlias(child, _) => child
+      case _ => expr
+    }
+
+    /** Extracts a [[Generator]] expression and any names assigned by aliases to their output. */
+    private object AliasedGenerator {
+      def unapply(e: Expression): Option[(Generator, Seq[String])] = e match {
+        case Alias(g: Generator, name) if g.resolved && g.elementSchema.length > 1 =>
+          // If not given the default names, and the TGF with multiple output columns
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Dumb question: what is TGF? Table Generating Function?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 4]
```diff
@@ -41,19 +41,17 @@ import org.apache.spark.sql.types._
...
+  override def dataType: DataType = ArrayType(elementSchema)
 
   override def foldable: Boolean = false
 
   override def nullable: Boolean = false
 
   /**
-   * The output element data types in structure of Seq[(DataType, Nullable)]
+   * The output element schema.
    * TODO we probably need to add more information like metadata etc.
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: MINOR: is this TODO still valid?

**Comment by hvanhovell** [comment]
LGTM. Left a few tiny tiny comments. Feel free to address those.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yea, and this comment is just copied from the original code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: I'm not sure, cc @davies @marmbrus

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: I did not know either

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Let's just remove it. This is why TODOs in code are often not very useful :)

**Comment by rxin** [comment]
Merging in master/branch-2.0.

**Pull request closed without merging**